internal/pkg/selinux/selinux.go

```diff
@@ -0,0 +1,99 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package selinux provides generic code for managing SELinux.
+package selinux
+
+import (
+	_ "embed"
+	"log"
+	"os"
+	"runtime"
+
+	"github.com/siderolabs/go-procfs/procfs"
+	"golang.org/x/sys/unix"
+
+	"github.com/siderolabs/talos/pkg/machinery/constants"
+)
+
+//go:embed policy/policy.33
+var policy []byte
+
+// IsEnabled checks if SELinux is enabled on the system by reading
+// the kernel command line. It returns true if SELinux is enabled,
+// otherwise it returns false. It also ensures we're not in a container.
+// By default SELinux is disabled.
+func IsEnabled() bool {
+	// TODO: resolve circular dependency with platform
+	if _, err := os.Stat("/usr/etc/in-container"); err == nil {
+		return false
+	}
+
+	val := procfs.ProcCmdline().Get(constants.KernelParamSELinux).First()
+
+	return val != nil && *val == "1"
+}
+
+// IsEnforcing checks if SELinux is enabled and the mode should be enforcing.
+// By default if SELinux is enabled we consider it to be permissive.
+func IsEnforcing() bool {
+	if !IsEnabled() {
+		return false
+	}
+
+	val := procfs.ProcCmdline().Get(constants.KernelParamSELinuxEnforcing).First()
+
+	return val != nil && *val == "1"
+}
+
+// SetLabel sets label for file or directory, following symlinks
+// It does not perform the operation in case SELinux is disabled or provided label is empty.
+func SetLabel(filename string, label string) error {
+	if label == "" {
+		return nil
+	}
+
+	if IsEnabled() {
+		if err := unix.Lsetxattr(filename, "security.selinux", []byte(label), 0); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Init initializes SELinux based on the configured mode.
+// It loads the policy and enforces it if necessary, and also prepares for launching machined.
+func Init() error {
+	if !IsEnabled() {
+		log.Println("selinux: disabled, not loading policy")
+
+		return nil
+	}
+
+	log.Println("selinux: loading policy")
+
+	if err := os.WriteFile("/sys/fs/selinux/load", policy, 0o777); err != nil {
+		return err
+	}
+
+	log.Println("selinux: policy loaded")
+
+	if IsEnforcing() {
+		log.Println("selinux: setting mode to enforcing")
+
+		if err := os.WriteFile("/sys/fs/selinux/enforce", []byte("1"), 0o777); err != nil {
+			return err
+		}
+	}
+
+	// Ensure we run in the same task to transition the init process into proper context
+	runtime.LockOSThread()
+
+	if err := os.WriteFile("/proc/thread-self/attr/exec", []byte(constants.SelinuxLabelMachined), 0o777); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
should we unlock the os thread here, or it's okay to do in init?

Reply:
It's certainly not ok here, since init then needs to exec from the same thread. Maybe locking should be better moved into switchroot.

Reply:
Dropped as I moved to typetransition'ing kernel_t -> init_t when executing init_exec_t